Security.md

```diff
@@ -0,0 +1,122 @@
+infinite precision to avoid wrapping and simplify bounds checking. In the future,
+support for [multiple linear memory sections](Modules.md#linear-memory-section) and
+[finer-grained memory operations](FutureFeatures.md#finer-grained-control-over-memory)
+(e.g. shared memory, page protection, large pages, etc.) will be implemented.
+
+[Traps](AstSemantics.md#traps) are used to immediately terminate execution and
+signal abnormal behavior to the execution environment. In a browser, this is
+represented as a JavaScript exception. Operations that can result in a trap
+include specifying an invalid index in any index space, performing an indirect
+function call with a mismatched signature, exceeding the maximum size of the
+protected call stack, and accessing [out-of-bounds](#Rationale.md#out-of-bounds)
+addresses in linear memory, as well as certain illegal arithmetic operations
+such as division or remainder by zero, and signed division overflow. In the
+future,
+[module-defined trap handlers](FutureFeatures.md#trappingor-non-trapping-strategies)
+will be supported.
+
+### Memory Safety
+Compared to traditional C/C++ programs, these semantics obviate certain classes
+of memory safety bugs in WebAssembly. Buffer overflows, which occur when data
+exceeds the boundaries of an object and accesses adjacent memory regions, cannot
+affect local or global variables stored in index space, they are fixed-size and
+addressed by index. Data stored in linear memory can overwrite adjacent objects,
+since bounds checking is performed at linear memory region granularity and is
+not context-sensitive. However, the presence of control-flow integrity and
+protected shadow call stacks prevents direct code injection attacks. Thus,
+common mitigations such as [data execution prevention][] (DEP) and
+[stack smashing protection][] (SSP) are not useful.
+
+Another common class of memory safety errors involves unsafe pointer usage and
+[undefined behavior](CandC++.md#undefined-behavior). This includes dereferencing
+pointers to unallocated memory (e.g. `NULL`), or freed memory allocations. In
+WebAssembly, the semantics of pointers have been eliminated for function calls
+and variables with fixed static scope, allowing references to invalid indexes in
+any index space to trigger a validation error at load time, or at worst a trap
+at runtime. Accesses to linear memory are bounds-checked at the region level,
+potentially resulting in a trap at runtime. These memory region(s) are isolated
+from the memory space of the runtime, and are set to zero by default unless
+otherwise initialized.
+
+Nevertheless, other classes of bugs are not obviated by the semantics of
+WebAssembly. Although attackers cannot perform direct code injection attacks,
+it is possible to hijack the control flow of a module using code reuse attacks
+(similar to [return-oriented programming][] (ROP)) against indirect calls with
+function-level granularity. Likewise, aside from
```

Comment:
What does "function-level" granularity refer to here? Is it saying that control flow through indirect calls can be hijacked to go to undesired functions but not in the middle of an instruction or an arbitrary instruction inside of a function?